crates/red_knot_python_semantic/src/types.rs

```diff
@@ -6188,13 +6205,8 @@ impl<'db> ConstructorCallError<'db> {
             }
         };
 
-        let report_new_error = |call_dunder_error: &CallDunderError<'db>| match call_dunder_error {
-            CallDunderError::MethodNotAvailable => {
-                // We are explicitly checking for `__new__` before attempting to call it,
-                // so this should never happen.
-                unreachable!("`__new__` method may not be called if missing");
-            }
-            CallDunderError::PossiblyUnbound(bindings) => {
+        let report_new_error = |error: &DunderNewCallError<'db>| match error {
```

Comment:
nit: I would name this variable

```suggestion
        let report_dunder_new_error = |error: &DunderNewCallError<'db>| match error {
```

to make clear that it's an error to do with the `__new__` method, not a "new error"